Indented string display

romainfd opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fixes https://github.com/Qiskit/qiskit-aqua/issues/1099
Current results
```
ComposedOp([
	SummedOp([
		ComposedOp([
			OperatorMeasurement(AbelianSummedOp([
	0.18093119978423136 * Z,
	-1.052373245772859 * I
])),
			     ┌───┐
q_0: ┤ H ├
     └───┘
		]),
		ComposedOp([
			OperatorMeasurement(AbelianSummedOp([
	0.18093119978423136 * Z,
	-2.104746491545718 * I
])),
			I
		])
	]),
	     ┌───┐
q_0: ┤ H ├
     └───┘
])
```

My current implementation (adding another display function passing the indentation level through its calls) is working for `ListOp` 
```
ComposedOp([
	SummedOp([
		X,
		Z
	])
])
```
but making it work for more cases requires a lot of changes including changes to `QuantumCircuit` to make `StateFn` display properly.
Did you have another implementation in mind?

### Details and comments

woodsp-ibm commented:
Hi, I had not really thought too much about the issue as it was raised quite recently. Here is a thought. How about a method that all the opflow entities can use if they wish in OperatorBase where I see you put something. Lets call it say `indent(lines: str) -> str` so given an string it will simply return an (one level) indented version back. I.e it will turn it into a list of strings at newline boundaries, prepend some indent to each string in the list, and the join them all back to a single string to return.

Now if any opflow entity is including other parts that it needs indenting, such as the string from a QuantumCircuit you can now do `indent(str(qc))` for example. If each opflow entity does its own indent of what it needs in its __str__ then it should all workout I think.

One concern with indenting in general might be that the output marches off to the right quite fast. Guess we'll have to see with how much indenting/nesting goes on in the most common sort of usage. Limiting the indent to say 2 or 4 spaces may still aid readability but keep it from marching to the right to fast. Since the indent would be in one place it would be easy enough to adjust. We could even set up OperatorBase to allow the indent level to be set, so as a user could override whatever default we choose.

Anyway hope this thought helps.

romainfd commented:
These are really interesting thoughts, thank you!
The `indent` approach is a great idea, I'll work on that! Now that you've suggesting it, it feels quite natural to pull the indentation from the callee to the caller to reduce changes downstream / on end dependencies and I wish I thought about that ^^
I'll add the centralized indentation + override capability as indentation could indeed be quite harmful to visualization...
Thank you for your good feedback. Not sure I'll have time to work on it this week but the issue said 
> This is a just requirement for nice-to-have.
so hopefully this is fine.

romainfd commented:
@woodsp-ibm thanks again for your previous comment. I worked on implementing this logic and it seems like having it in `ListOp` is enough as I felt like all Operators with nested elements inherit from it.

As of now, I put `indentation` as a parameter to `__str__` but I don't think this is an easy way to set it up because it doesn't allow global setup. I couldn't find any class variables (`ListOp.indentation`) or global variables (`INDENTATION`) for any other operators so wasn't sure about your preferred / consistent way to go.

woodsp-ibm commented:
@romainfd The output certainly looks a lot nicer now than before. So none of the primitive operators need any form of indent, or whatever indent level they might use for their own formatting of content is fixed and looks ok I guess. The indenting you are adding is just around the listops then to ensure each element comes out indented. I had imagined some class or module level setting, maybe in OperatorBase class or the module its in, that could be set to control the indent. At that level it could be read by any of the opflow that wanted to allow control over the indenting - maybe at present only listop would use it. The operator_globals file might also be a place for it - maybe a better place - it has some global flag in there already around precision. Maybe a fixed level of indent is good enough, and some global control is not needed - the output seems way more readable already.

woodsp-ibm commented on qiskit/aqua/operators/list_ops/list_op.py:
```diff
@@ -297,9 +297,16 @@ def log_i(self, massive: bool = False) -> OperatorBase:
+    @staticmethod
+    def _indent(lines: str, indentation: str = "  ") -> str:
+        """ Indented representation to allow pretty representation of nested operators. """
+        return indentation + lines.replace("\n", f"\n{indentation}").rstrip(indentation)
```
```suggestion
        return indentation + lines.replace("\n", "\n{}.format(indentation)").rstrip(indentation)
```
Since we still support Python 3.5 the use of `f"` style formatted strings got raised as a mypy issue

woodsp-ibm replied:
Can you please add unit test case(s) that create some operator(s) and check the output string meets expectation.

woodsp-ibm commented:
Can you please add unit test case(s) that create some operator(s) and check the output string meets expectation.

romainfd commented:
Hey, thanks for the good comments. I am on holidays this week but will work on that when I come back!

romainfd commented:
> The operator_globals file might also be a place for it 

Adding it there caused problems while importing it in Operators (I think it creates import cycles) so I moved it to `OperatorBase`. With this approach, it can also be updated for a subclass only (doing `ListOp.INDENTATION = '\t'`) so I think this is quite good.

As asked, I also added a few unittests. 

Let me know if there is something else I can improve.

romainfd commented:
No worries, my pleasure!
I added a comment based on what you suggested. 
I marked it ready for review but indeed forgot to remove the [WIP], sorry!
